packages/admin-graphql-api-utilities/src/index.ts

```diff
@@ -43,26 +50,51 @@ export function parseGidWithParams(gid: string): ParsedGid {
   throw new Error(`Invalid gid: ${gid}`);
 }
 
-export function composeGidFactory(namescape: string) {
-  return function composeGid(
-    key: string,
+export function composeGidFactory<N extends string>(namespace: N) {
+  return function composeGid<T extends string>(
+    key: T,
     id: number | string,
     params: {[key: string]: string} = {},
-  ): string {
-    const gid = `gid://${namescape}/${key}/${id}`;
+  ): Gid<N, T> {
+    const gid = `gid://${namespace}/${key}/${id}`;
     const paramKeys = Object.keys(params);
 
     if (paramKeys.length === 0) {
-      return gid;
+      return gid as Gid<N, T>;
     }
 
     const paramString = new URLSearchParams(params).toString();
-    return `${gid}?${paramString}`;
+    return `${gid}?${paramString}` as Gid<N, T>;
   };
 }
 
 export const composeGid = composeGidFactory('shopify');
 
+export function isGidFactory<N extends string>(namespace: N) {
+  return function isGid<T extends string = unknown>(
```

Comment:
Is the reason for defaulting this to `unknown` so that if you don't provide a key when calling `isGid` you don't have to provide this generic type? I played around to see if it'd be possible to get around that with function overloading (like one version of `isGid` without the key parameter and another with the key parameter but it's non-optional) ... trying to think of how to get rid of the `unknown` since that's tripping this up.

Reply:
Yeah that was the intention, but we could just make this default to `string`. Would that work?